File: src/main/java/com/ata1635/cardgame/core/games/Blackjack.java
package com.ata1635.cardgame.core.games;

import com.ata1635.cardgame.core.model.Card;
import com.ata1635.cardgame.core.service.Deck;

import java.util.ArrayList;
import java.util.List;

public class Blackjack {

    private final Deck deck;
    private Card hiddenCard;

    private ArrayList<Card> dealerHand;
    private ArrayList<Card> playerHand;

    private int dealerSum;
    private int playerSum;
    private int dealerAceCount; //devalue the Ace from 11 to 1
    private int playerAceCount;

    public Blackjack() {
        deck = new Deck();
        deck.shuffleDeck();
        startGame();
    }

    //constructor for testing
    public Blackjack(Deck custonDeck) {
        this.deck = custonDeck;
        startGame();
    }

    public void startGame() {
        //dealer
        dealerHand = new ArrayList<Card>();
        dealerSum = 0;
        dealerAceCount = 0;

        hiddenCard = deck.drawCard();
        dealerSum += hiddenCard.getValue();
        dealerAceCount += hiddenCard.isAce() ? 1 : 0;

        Card card = deck.drawCard(); //dealer only draws one card
        dealerSum += card.getValue();
        dealerAceCount += card.isAce() ? 1 : 0;
        dealerHand.add(card);
        adjustDealerAce();

        //player
        playerHand = new ArrayList<Card>();
        playerSum = 0;
        playerAceCount = 0;

        for (int i = 0; i < 2; i++) { //player draws 2 cards at start
            card = deck.drawCard();
            playerSum += card.getValue();
            playerAceCount += card.isAce() ? 1 : 0;
            playerHand.add(card);
        }
        adjustPlayerAce();
        printHands();
    }

    public Card playerHit() {
        Card card = deck.drawCard();
        playerHand.add(card);
        playerSum += card.getValue();
        playerAceCount += card.isAce() ? 1 : 0;
        adjustPlayerAce();
        printHands();
        return card;
    }

    public Card dealerHit() {
        Card card = deck.drawCard();
        dealerHand.add(card);
        dealerSum += card.getValue();
        dealerAceCount += card.isAce() ? 1 : 0;
        adjustDealerAce();
        printHands();
        return card;
    }

    private void adjustPlayerAce() {
        while (playerSum > 21 && playerAceCount > 0) {
            playerSum -= 10;
            playerAceCount--;
        }
    }

    private void adjustDealerAce() {
        while (dealerSum > 21 && dealerAceCount > 0) {
            dealerSum -= 10;
            dealerAceCount--;
        }
    }

    public boolean isPlayerBust() {
        return playerSum > 21;
    }

    public void dealerFinalHits() {
        while (dealerSum < 17) {
            Card card = deck.drawCard();
            dealerSum += card.getValue();
            dealerAceCount += card.isAce() ? 1 : 0;
            dealerHand.add(card);
            printHands();
        }
    }

    public void printHands() {
        System.out.println("DEALER:\n" + hiddenCard + "\n"
                + dealerHand + "\n" + dealerSum + "\n" + dealerAceCount + "\n");
        System.out.println("PLAYER:\n" + playerHand + "\n" + playerSum + "\n" + playerAceCount);
        System.out.println("\n ---------- \n");
    }

    public List<Card> getDealerHand() { return dealerHand; }
    public List<Card> getPlayerHand() { return playerHand; }
    public Card getHiddenCard() { return hiddenCard; }
    public int getDealerAceCount() { return dealerAceCount; }
    public int getPlayerAceCount() { return playerAceCount; }
    public int getDealerSum() { return dealerSum; }
    public int getPlayerSum() { return playerSum; }

}

File: src/js/core/games/Blackjack.js
import { Deck } from '../service/Deck.js';

export class Blackjack {
    constructor(customDeck = null) {
        this.deck = customDeck || new Deck();
        this.deck.shuffleDeck();
        this.startGame();
    }

    startGame() {
        // Dealer
        this.dealerHand = [];
        this.dealerSum = 0;
        this.dealerAceCount = 0;

        this.hiddenCard = this.deck.drawCard();
        this.dealerSum += this.hiddenCard.getValue();
        this.dealerAceCount += this.hiddenCard.isAce() ? 1 : 0;

        const card = this.deck.drawCard(); // Dealer draws 1 card
        this.dealerSum += card.getValue();
        this.dealerAceCount += card.isAce() ? 1 : 0;
        this.dealerHand.push(card);
        this.adjustDealerAce();

        // Player
        this.playerHand = [];
        this.playerSum = 0;
        this.playerAceCount = 0;

        for (let i = 0; i < 2; i++) {
            const playerCard = this.deck.drawCard();
            this.playerSum += playerCard.getValue();
            this.playerAceCount += playerCard.isAce() ? 1 : 0;
            this.playerHand.push(playerCard);
        }
        this.adjustPlayerAce();
        this.printHands();
    }

    playerHit() {
        const card = this.deck.drawCard();
        this.playerHand.push(card);
        this.playerSum += card.getValue();
        this.playerAceCount += card.isAce() ? 1 : 0;
        this.adjustPlayerAce();
        this.printHands();
        return card;
    }

    dealerHit() {
        const card = this.deck.drawCard();
        this.dealerHand.push(card);
        this.dealerSum += card.getValue();
        this.dealerAceCount += card.isAce() ? 1 : 0;
        this.adjustDealerAce();
        this.printHands();
        return card;
    }

    adjustPlayerAce() {
        while (this.playerSum > 21 && this.playerAceCount > 0) {
            this.playerSum -= 10;
            this.playerAceCount--;
        }
    }

    adjustDealerAce() {
        while (this.dealerSum > 21 && this.dealerAceCount > 0) {
            this.dealerSum -= 10;
            this.dealerAceCount--;
        }
    }

    isPlayerBust() {
        return this.playerSum > 21;
    }

    dealerFinalHits() {
        while (this.dealerSum < 17) {
            const card = this.deck.drawCard();
            this.dealerSum += card.getValue();
            this.dealerAceCount += card.isAce() ? 1 : 0;
            this.dealerHand.push(card);
            this.printHands();
        }
    }

    printHands() {
        console.log("DEALER:\n", this.hiddenCard.toString(), "\n", this.dealerHand.map(c => c.toString()), "\nSum:", this.dealerSum, "Aces:", this.dealerAceCount);
        console.log("PLAYER:\n", this.playerHand.map(c => c.toString()), "\nSum:", this.playerSum, "Aces:", this.playerAceCount);
        console.log("\n ---------- \n");
    }

    getDealerHand() {
        return this.dealerHand;
    }

    getPlayerHand() {
        return this.playerHand;
    }

    getHiddenCard() {
        return this.hiddenCard;
    }

    getDealerAceCount() {
        return this.dealerAceCount;
    }

    getPlayerAceCount() {
        return this.playerAceCount;
    }

    getDealerSum() {
        return this.dealerSum;
    }

    getPlayerSum() {
        return this.playerSum;
    }
}